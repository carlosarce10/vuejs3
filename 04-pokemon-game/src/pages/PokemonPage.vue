<template>
  <h1 v-if="!pokemon">Espere por favor...</h1>
  <div v-else>
    <h1>¿Qué pokemon es?</h1>
    <PokemonPicture
      data-testid="pokemonPicture"
      :pokemonId="pokemon.id"
      :showPokemon="showPokemon"
    />
    <PokemonOptions
      data-testid="pokemonOptions"
      :pokemons="pokemonArr"
      @selection="checkAnswer"
    />

    <template class="fade-in" v-if="showAnswer">
      <h2>{{ message }}</h2>
      <button @click="newGame">Nuevo juego</button>
    </template>
  </div>
</template>

<script>
import PokemonOptions from '@/components/PokemonOptions.vue';
import PokemonPicture from '@/components/PokemonPicture.vue';
import getPokemonOptions from '@/helpers/getPokemonOptions.js';

export default {
  components: {
    PokemonOptions,
    PokemonPicture,
  },
  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: '',
    };
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions();

      const rndId = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemonArr[rndId];
    },
    checkAnswer(pokemonId) {
      this.showPokemon = true;
      this.showAnswer = true;

      pokemonId == this.pokemon.id
        ? (this.message = 'Correcto')
        : (this.message = `Fallaste, el pokemon es ${this.pokemon.name}`);
    },
    newGame() {
      this.mixPokemonArray();
      this.showPokemon = false;
      this.showAnswer = false;
      this.pokemonArr = [];
      this.pokemon = null;
    },
  },
  mounted() {
    this.mixPokemonArray();
  },
};
</script>

<style></style>
