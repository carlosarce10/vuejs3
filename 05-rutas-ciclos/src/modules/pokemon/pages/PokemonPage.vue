<template>
  <h1>
    Pokemon: <span>#{{ id }}</span>
  </h1>
  <div v-if="pokemon">
    <img :src="pokemon.sprites.front_default" :alt="pokemon.name" />
  </div>
</template>

<script>
export default {
  props: {
    id: {
      type: Number,
      required: true,
    },
  },
  data() {
    return {
      // id: null,
      pokemon: null,
    };
  },
  created() {
    /* console.log(this.$route);
    this.id = this.$route.params.id; */

    this.getPokemon();
  },
  methods: {
    async getPokemon() {
      try {
        console.log('entro');
        const pokemon = await fetch(
          `https://pokeapi.co/api/v2/pokemon/${this.id}`
        ).then((response) => response.json());
        this.pokemon = pokemon;
      } catch (error) {
        this.$router.push('/');
        console.log('No existe el pokemon', error);
      }
    },
  },
  watch: {
    id() {
      this.getPokemon();
    },
  },
};
</script>
