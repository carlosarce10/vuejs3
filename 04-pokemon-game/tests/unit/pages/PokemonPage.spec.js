import PokemonPage from '@/pages/PokemonPage';
import { shallowMount, mount } from '@vue/test-utils';
import { pokemons } from '../mocks/pokemos.mock';

describe('PokemonPage component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage);
  });

  test('debe de llamar mixPokemonArray al montar', () => {
    const mixPokemonArray = jest.spyOn(PokemonPage.methods, 'mixPokemonArray');
    wrapper = shallowMount(PokemonPage);

    expect(mixPokemonArray).toHaveBeenCalled();
  });

  test('debe de hacer match con el snapshot cuando carga los pokemos', () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: '',
        };
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  test('debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: '',
        };
      },
    });

    const picture = wrapper.find('[data-testid="pokemonPicture"]');
    const options = wrapper.find('[data-testid="pokemonOptions"]');

    expect(picture.exists()).toBeTruthy();
    expect(options.exists()).toBeTruthy();

    expect(picture.attributes('pokemonid')).toBe('1');
    expect(options.attributes('pokemons')).toBeTruthy();
  });

  test('pruebas con checkAnswer', async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: '',
        };
      },
    });
    await wrapper.vm.checkAnswer(1);

    expect(wrapper.find('h2').exists()).toBeTruthy();
    expect(wrapper.vm.showPokemon).toBe(true);
    expect(wrapper.find('h2').text()).toBe('Andas duro, kbron');

    await wrapper.vm.checkAnswer(10);
    expect(wrapper.vm.message).toBe(
      `No le sabes, el pokemon es ${pokemons[0].name}`
    );
  });
});
