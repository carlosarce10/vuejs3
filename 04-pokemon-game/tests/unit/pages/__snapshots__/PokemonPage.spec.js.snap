// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component debe de hacer match con el snapshot cuando carga los pokemos 1`] = `
<div>
  <h1>¿Qué pokemon es?</h1>
  <pokemon-picture-stub pokemonid="1" showpokemon="false" data-testid="pokemonPicture"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]" data-testid="pokemonOptions"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
